File: app/src/main/java/com/ndhunju/dailyjournal/controller/service/GoogleSignInHelper.java
package com.ndhunju.dailyjournal.controller.service;

import android.content.Context;

import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.GoogleApiAvailability;
import com.google.android.gms.common.api.Scope;
import com.google.api.client.extensions.android.http.AndroidHttp;
import com.google.api.client.extensions.android.json.AndroidJsonFactory;
import com.google.api.client.googleapis.extensions.android.gms.auth.GoogleAccountCredential;
import com.google.api.services.drive.Drive;
import com.google.api.services.drive.DriveScopes;
import com.google.common.collect.Sets;
import com.ndhunju.dailyjournal.R;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

import androidx.annotation.NonNull;
import androidx.core.util.Pair;

import static com.ndhunju.dailyjournal.controller.service.DriveServiceHelper.OPERATION_STATUS_FAIL;

/**
 * Helper class that groups relevant objects like {@link Scope} and provides helper methods
 */
public class GoogleSignInHelper {

    public static final GoogleSignInHelper INSTANCE = new GoogleSignInHelper();

    public static GoogleSignInHelper get() {
        return INSTANCE;
    }

    Scope[] requiredScopes = {
            // See https://github.com/vaquarkhan/Scoot-mobile-app/blob/master/ScootAppSource/com/google/android/gms/auth/api/signin/GoogleSignInOptions.java
//            GoogleSignInOptions.zat, // openId
//            GoogleSignInOptions.zar, // profile
            // Without email, gives "IllegalArgumentException: the name must not be empty: null"
            /** See {@link GoogleSignInOptions.Builder#requestEmail()} to know which var is for email */
            GoogleSignInOptions.zab, // email
            new Scope(DriveScopes.DRIVE_FILE)
    };

    private GoogleSignInHelper() {}

    public GoogleSignInOptions buildGoogleSigInOptions() {
        GoogleSignInOptions.Builder builder = new GoogleSignInOptions.Builder();

        for (Scope scope: requiredScopes) {
            builder.requestScopes(scope);
        }

        return builder.build();
    }

    public GoogleSignInClient getGoogleSigInClient(Context context) {
        return GoogleSignIn.getClient(
                context,
                GoogleSignInHelper.get().buildGoogleSigInOptions()
        );
    }

    @NonNull
    public Drive signInToGoogleDrive(GoogleSignInAccount googleSignInAccount, Context context) {
        // Use the authenticated account to sign in to the Drive service.
        GoogleAccountCredential credential = GoogleAccountCredential.usingOAuth2(
                context,
                requiredScopesAsStringList()
        );

        credential.setSelectedAccount(googleSignInAccount.getAccount());

        return new Drive.Builder(
                AndroidHttp.newCompatibleTransport(),
                new AndroidJsonFactory(),
                credential
        ).setApplicationName(context.getString(R.string.app_name)).build();
    }

    public Pair<GoogleSignInAccount, Integer> getLastSignedInAccountAndConnectionResult(
            Context context
    ) {

        int googleServiceStatus = GoogleApiAvailability.getInstance()
                .isGooglePlayServicesAvailable(context);

        switch (googleServiceStatus) {
            case ConnectionResult.SERVICE_MISSING:
            case ConnectionResult.SERVICE_VERSION_UPDATE_REQUIRED:
            case ConnectionResult.API_UNAVAILABLE:
            case ConnectionResult.SERVICE_DISABLED:
                return Pair.create(null, googleServiceStatus);
            case ConnectionResult.SUCCESS:
                GoogleSignInAccount signInAccount = GoogleSignIn.getLastSignedInAccount(context);
                if (signInAccount != null
                        && !signInAccount.isExpired()
                        && (signInAccount.getGrantedScopes()
                        .containsAll(requiredScopesAsSet()))
                        && (DriveServiceHelper.getLastOperationStatus(context)
                        != OPERATION_STATUS_FAIL)) {
                    return Pair.create(signInAccount, googleServiceStatus);
                }
        }

        return Pair.create(null, googleServiceStatus);
    }

    public Set<Scope> requiredScopesAsSet() {
        return Sets.newHashSet(requiredScopes);
    }

    public List<String> requiredScopesAsStringList() {
        List<String> scopes = new ArrayList<>();
        for (Scope scope: requiredScopes) {
            scopes.add(scope.toString());
        }

        return scopes;
    }
}

File: app/src/main/java/com/ndhunju/dailyjournal/controller/service/GoogleSignInHelper.kt
package com.ndhunju.dailyjournal.controller.service

import android.content.Context
import android.util.Log
import androidx.annotation.NonNull
import androidx.annotation.Nullable
import androidx.core.util.Pair
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import com.google.android.gms.auth.api.signin.GoogleSignInClient
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.GoogleApiAvailability
import com.google.android.gms.common.api.Scope
import com.google.api.client.googleapis.extensions.android.gms.auth.GoogleAccountCredential
import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport
import com.google.api.client.json.gson.GsonFactory
import com.google.api.services.drive.Drive
import com.google.api.services.drive.DriveScopes
import com.google.common.collect.Sets
import com.ndhunju.dailyjournal.R

/**
 * Helper class that groups relevant objects like [Scope] and provides helper methods
 */
object GoogleSignInHelper {

    const val TAG = "GoogleSignInHelper"

    private val requiredScopes = setOf(Scope(DriveScopes.DRIVE_FILE))

    fun buildGoogleSigInOptions(): GoogleSignInOptions {
        return GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
            .requestEmail()
            .requestScopes(Scope(DriveScopes.DRIVE_FILE))
            .build()
    }

    fun getGoogleSigInClient(context: Context): GoogleSignInClient {
        return GoogleSignIn.getClient(context, buildGoogleSigInOptions())
    }

    fun signInToGoogleDrive(
        @Nullable googleSignInAccount: GoogleSignInAccount?,
        context: Context
    ): Drive? {
        googleSignInAccount ?: return null

        val credential = GoogleAccountCredential.usingOAuth2(
            context,
            listOf(DriveScopes.DRIVE_FILE)
        )

        credential.selectedAccount = googleSignInAccount.account
        return try {
            Drive.Builder(
                GoogleNetHttpTransport.newTrustedTransport(),
                GsonFactory.getDefaultInstance(),
                credential
            )
                .setApplicationName(context.getString(R.string.app_name))
                .build()
        } catch (e: Exception) {
            Log.e(TAG, "signInToGoogleDrive: ", e)
            null
        }
    }

    @NonNull
    fun getLastSignedInAccountAndConnectionResult(context: Context): Pair<GoogleSignInAccount?, Int> {
        val googleServiceStatus = GoogleApiAvailability.getInstance()
            .isGooglePlayServicesAvailable(context)

        return when (googleServiceStatus) {
            ConnectionResult.SERVICE_MISSING,
            ConnectionResult.SERVICE_VERSION_UPDATE_REQUIRED,
            ConnectionResult.API_UNAVAILABLE,
            ConnectionResult.SERVICE_DISABLED -> Pair(null, googleServiceStatus)

            ConnectionResult.SUCCESS -> {
                val signInAccount = GoogleSignIn.getLastSignedInAccount(context)
                if (signInAccount != null
                    && !signInAccount.isExpired
                    && requiredScopes.all { signInAccount.grantedScopes.contains(it) }
                    && DriveServiceHelper.getLastOperationStatus(context) != DriveServiceHelper.OPERATION_STATUS_FAIL
                ) {
                    Pair(signInAccount, googleServiceStatus)
                } else {
                    Pair(null, googleServiceStatus)
                }
            }

            else -> Pair(null, googleServiceStatus)
        }
    }

    fun requiredScopesAsSet(): Set<Scope> = Sets.newHashSet(*requiredScopes.toTypedArray())

    fun requiredScopesAsStringList(): List<String> = requiredScopes.map { it.toString() }
}
